plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergNodePartitioningProvider.java

```diff
@@ -26,22 +28,36 @@
 import javax.inject.Inject;
 
 import java.util.List;
+import java.util.Optional;
 
 import static io.trino.plugin.iceberg.IcebergUtil.schemaFromHandles;
 import static io.trino.plugin.iceberg.PartitionFields.parsePartitionFields;
+import static io.trino.spi.connector.ConnectorBucketNodeMap.createBucketNodeMap;
 import static java.util.Objects.requireNonNull;
 
 public class IcebergNodePartitioningProvider
         implements ConnectorNodePartitioningProvider
 {
+    private final NodeManager nodeManager;
     private final TypeOperators typeOperators;
 
     @Inject
-    public IcebergNodePartitioningProvider(TypeManager typeManager)
+    public IcebergNodePartitioningProvider(NodeManager nodeManager, TypeManager typeManager)
     {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
         this.typeOperators = requireNonNull(typeManager, "typeManager is null").getTypeOperators();
     }
 
+    @Override
+    public Optional<ConnectorBucketNodeMap> getBucketNodeMapping(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
+    {
+        if (partitioningHandle instanceof IcebergUpdateHandle) {
+            return Optional.empty();
+        }
+
+        return Optional.of(createBucketNodeMap(nodeManager.getRequiredWorkerNodes().size()));
```

Comment:
I think I don't understand the change.
Is `ConnectorNodePartitioningProvider.getBucketNodeMapping` mandatory to implement?
@electrum 's 32079253491a35b453d0034322701639e2a2f9d2 (part of https://github.com/trinodb/trino/pull/7933) suggests it should be optional to implement this method.

If it's optional, do we have a bug in the engine, which manifests only when this method is not implemented?
If so, shouldn't we have a fix in the engine?

Reply:
This should be a bug in the engine. I believe this implementation will break MERGE.

Reply:
This causes MERGE to fail with

> Insert and update layout have mismatched BucketNodeMap

Which is why we made it optional to implement this method. We need to track down why this fails. None of the existing integration tests caught this case.

Reply:
@electrum Can you take over the issue?

Reply:
Sure, I can look at this. Thanks for writing the test, it's helpful.